contracts/BitGuildMarketplace.sol

```diff
@@ -0,0 +1,355 @@
+    //new version with operator
+    bytes4 constant ERC721_RECEIVED = 0x150b7a02;
+
+    // Predefined PLAT token
+    // BitGuildToken public token = BitGuildToken(0x7E43581b19ab509BCF9397a2eFd1ab10233f27dE); // Main Net
+    BitGuildToken public PLAT = BitGuildToken(0x0F2698b7605fE937933538387b3d6Fec9211477d); // Rinkeby
+
+    uint public defaultExpiry = 7 days;   // default expiry 7 days
+    // TODO: add function to modify this length 
+
+    uint public currentItemId = 1; // auto-increment
+
+    mapping(uint256 => address) public itemIdToContract;       // game contract the item belongs to
+    mapping(uint256 => address) public itemIdToSeller;         // seller of the item. 0x0 if sold/delisted
+    mapping(uint256 => uint256) public itemIdToTokenId;        // tokenId for this item
+    mapping(uint256 => uint256) public itemIdToPLAT;           // price in PLAT for this item
+    mapping(uint256 => uint256) public itemIdToETH;            // price in ETH for this item
+    mapping(uint256 => uint256) public itemIdToCreatedAt;      // timestamp when item is listed
+
+    event ItemDeposited(address _seller, uint _itemId, uint _createdAt, uint expiry);
+    event ItemSold(address _buyer, address _seller, uint _itemId, uint _ETH, uint _PLAT);
+    event ItemWithdrawn(address _seller, uint _itemId, uint _withdrawnAt);
+
+    modifier isTokenOwner(address _contract, uint _tokenId) {
+        ERC721Interface gameToken = ERC721Interface(_contract);
+        require(
+            gameToken.ownerOf(_tokenId) == msg.sender,
+            "Seller does not own the token."
+        );
+        _;
+    }
+
+    modifier isItemSeller(uint _itemId) {
+        require(
+            itemIdToSeller[_itemId] == msg.sender,
+            "Not item seller."
+        );
+        _;
+    }
+
+    modifier isPriceValid(uint _price) {
+        require(
+            _price > 0,
+            "Invalid price."
+        );
+        _;
+    }
+
+    modifier doPricesMatchETH(uint _itemId) {
+        require(
+            itemIdToETH[_itemId] == msg.value,
+            "Mismatched ETH prices."
+        );
+        _;
+    }
+
+    modifier doPricesMatchPLAT(uint _itemId, uint _price) {
+        require(
+            itemIdToPLAT[_itemId] == _price,
+            "Mismatched PLAT prices."
+        );
+        _;
+    }
+
+    modifier isPLAT() {
+        // Make sure it is sent from BitGuildToken
+        require(
+            msg.sender == address(PLAT),
+            "Unauthorized PLAT contract address."
+        );
+        _;
+    }
+
+    // @dev fall back function
+    function () external payable {
+        revert();
+    }
+
+    // @dev return all available item ids
+    function getAllItemIds() public view returns(uint[]) {
+        // @dev Push is not supported for memory. storage costs more gas. Use manual counter instead
+        uint[] memory tokens;
+        uint counter = 0;
+        for (uint i = 1; i <= currentItemId; i++) {
```

Comment:
This is crazy and will die on any significant number of items. You cannot have a loop over all items.